### PR TITLE
Autoscaling

### PR DESCRIPTION
**Pull request by s373nZ**
Replace single EC2 instance architecture with Auto Scaling Group behind Application Load Balancer for horizontal scaling during traffic spikes.

Key changes:
- Packer configuration for building base AMIs with Ansible
- Split Ansible playbook into base (AMI) and runtime (boot) phases
- ALB with ACM certificate, sticky sessions for WebSocket support
- ASG with mixed instances policy (On-Demand base + Spot for scaling)
- Target tracking scaling policies (CPU, request count)
- SSM Parameter Store for secrets management
- Enhanced health check with database connectivity verification
- Improved graceful shutdown with timeout handling
- CI/CD pipeline with AMI change detection and instance refresh

This is a biiiiig change, but is now working for me on my fork's staging environment. I've got it deployed through the `develop` pipeline and have been running the performance tests in #103 for `make test-perf-load` against it. It's autoscaling successfully and have been working through tuning it a bit.

Currently it takes about 2.5 minutes to scale in new instances from the `t3.small`. After scaling up to about 10 instances during the load test, it take about 30 minutes to scale back down to one instance. Need to run the test again since I've done a little bit more tuning, but we did have a period of failed requests in the 2-3 minutes while scaling up from the first instance. I've lowered the scaling thresholds, but the best solution would be to start with a bigger base instance size which wouldn't get overwhelmed as easily, but that would cost more. Probably a good idea to bump up the instance size before known classes or scheduled events.

An important note, I've left the original `ansible/playbook.yaml` file we are using to deploy to a single EC2 instance. This could be easily re-used to configure another standalone server on EC2 or another hosting platform if cost or operational complexity trade-offs become too large.

In terms of review - auditing the documentation blocks to be sure the approach is clear could be a good start. It's also very helpful to see it in action -- merging to `develop` to try to deploy it and run the tests, or schedule a demo call.

TODO:
- ~(Even) more code review~
- ~Documentation review and update~
- ~Additional load testing?~
- Run the E2E test suite against this infrastructure -- it was "mostly" working, but failed a few tests; maybe data integrity related?

### CODE REVIEW

**Comment by s373nZ** [comment]
A few more commits to:
- Rename and expose the `fallback_instance_types` variable through GitHub Actions to allow specification of the ASG backup instance types.
- Use a bigger image to build the AMIs, speeding up the process a little.
- Enable the ASG group metrics by default.

I also ran the tests on a `c6a.xlarge` instance today out of curiousity, and it performed very well. Not perfect, but very well.

[test-summary-c6a.xlarge.log](https://github.com/user-attachments/files/27902098/test-summary-c6a.xlarge.log)

File: ansible/README.md [hunk 1]
```diff
@@ -1,38 +1,57 @@
...
+Run automatically on each server at first boot via the instance's startup script. Checks out
+the correct application version from Git, installs packages, writes the environment config
+(fetched from AWS SSM), builds the app, runs database migrations, and starts the service.
+**Not intended for manual use.**
+
+### `playbook.yaml` — Direct VPS deployment
+
+The original single-server playbook. Retained as an escape hatch if the AWS auto-scaling
+setup ever becomes too costly or operationally complex to justify — for example, when
+migrating to a VPS provider like Hetzner or DigitalOcean. It handles the full setup in one
```
Comment (jrakibi) [inline ansible/README.md]: nit: not sure if we should mention this as an "escape hatch" to Hetzner/DigitalOcean. Looking at playbook.yaml, I guess we're still relying on AWS (Cloudwatch / Route53) unless we modify the file.

File: packer/ami.pkr.hcl [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+variable "aws_region" {
+  type        = string
+  default     = "us-west-2"
+  description = "AWS region to build the AMI in"
+}
+
+variable "base_ami" {
+  type        = string
+  default     = "ami-0786adace1541ca80"
+  description = "Base Ubuntu 24.04 AMI to build from"
```
Comment (jrakibi) [inline packer/ami.pkr.hcl:23]: question: is the AMI linked to aws_region? I mean if we ever change aws_region to something else instead of us-west-2, this AMI `ami-0786adace1541ca80` won't work anymore and we'd have to update it?

if that's the case, not sure if we should make it an env variable cause the same ID is hard coded in 3 places. (Although I don't see any reason we'd change aws_region to smth else, just wanna mention this)

File: packer/variables.pkrvars.hcl [hunk 3]
```diff
@@ -0,0 +1,12 @@
+# Packer variables for AMI builds
+# Override defaults for production/staging builds
+
+aws_region      = "us-west-2"
+instance_type   = "c6a.xlarge"
```
Comment (satsie) [inline packer/variables.pkrvars.hcl:5]: This means that the AMI is c6a.xlarge, right? Does that mean the instance type is also c6a.xlarge?

File: packer/variables.pkrvars.hcl [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+ami_name_prefix = "saving-satoshi-backend"
+
+# Base AMI: Ubuntu 24.04 LTS in us-west-2
```
Comment (satsie) [inline packer/variables.pkrvars.hcl:8]: After some time we'll probably want to update to the Ubuntu 26.04 LTS. Is this the only place where the change will need to be made?

File: ansible/saving-satoshi-backend-alb.nginx.j2 [hunk 5]
```diff
@@ -0,0 +1,42 @@
+# Nginx configuration for ALB deployment
+# SSL termination is handled by the Application Load Balancer
+# This configuration only handles HTTP (port 80) traffic from ALB
```
Comment (satsie) [inline ansible/saving-satoshi-backend-alb.nginx.j2:3]: Is it port 80 because the ALB is the only thing that can reach NGINX? Because it's in a VPC?

File: ansible/playbook-runtime.yaml [hunk 6]
```diff
@@ -0,0 +1,120 @@
+---
+# Runtime playbook executed at instance boot via user-data
+# Deploys application code and starts services
+# Requires environment variables: APP_VERSION, APP_ENV_CONTENT, DATABASE_URL
```
Comment (satsie) [inline ansible/playbook-runtime.yaml:4]: I don't see `APP_VERSION` and `APP_ENV_CONTENT` in the Github Secrets and Variables section so I just want to confirm that `APP_ENV_CONTENT` is omitted from the other parts of the documentation where the variables are listed (like `DEPLOYMENT.md`) because this is automatically configured by `terraform/templates/user-data.sh`. Is the same for `APP_VERSION` because it's set by `ci.yml`?

File: ansible/playbook-runtime.yaml [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        recurse: true
+      loop:
+        - "{{ app_root }}/dist"
+        - "{{ app_root }}/logs"
+        - /var/log/saving-satoshi-backend
+
+    # ============================================================
+    # Docker Base Images
+    # ============================================================
+    - name: Build Docker base images for REPL
```
Comment (satsie) [inline ansible/playbook-runtime.yaml:87]: I thought `playbook-base.yaml` built the base images?

File: .env.example [hunk 8]
```diff
@@ -3,4 +3,10 @@ SECRET=12345678
...
 DATABASE_URL="postgresql://postgres:postgres@localhost:5432/postgres?schema=public"
-CONTAINER_TERMINATION_SCHEDULE= 10 # Scheduled Time to clear your container in minutes 1<=59
\ No newline at end of file
+CONTAINER_TERMINATION_SCHEDULE=10 # Scheduled Time to clear your container in minutes 1<=59
+
+# Database connection pool configuration for autoscaling
+# With multiple instances, configure connection_limit in DATABASE_URL to prevent pool exhaustion
+# Example: DATABASE_URL="postgresql://...?connection_limit=8&pool_timeout=30"
+# Recommended: (max_db_connections / max_instances) with some buffer
+# Default Prisma pool size is 5 connections
```
Comment (satsie) [inline .env.example:12]: Thanks for making note of this. I'm not sure if we set the connection pool to have more connections anywhere.

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -1,36 +1,27 @@
...
-#       REPL container timeouts.
-#   - PORT (optional): The application config value for the server listening port.
-#   - WHITELIST (optional): The application config value as a comma-separated list defining
-#       which allowed Origin headers are allowed to access the API.
+# Variables (Settings > Secrets and Variables > Actions > Variables):
+#   - AWS_EC2_INSTANCE_TYPE (optional): Primary instance type for ASG (default: t2.small)
+#   - AWS_EC2_KEY_PAIR_NAME: Name of existing EC2 key pair for SSH access
+#   - AWS_ROLE_ARN_TO_ASSUME: OIDC role ARN from `terraform/base` output
+#   - AWS_REGION: AWS region for deployment
+#   - AWS_ROUTE53_HOSTED_ZONE_ID: Hosted zone ID from `terraform/base` output
```
Comment (satsie) [inline .github/workflows/ci.yml]: This part of the description got removed but I think it could be helpful to have in the future:

> This value should be the output `dns_zone_id` from running the Terraform baseline infrastructure in `terraform/base`.

File: .github/workflows/ci.yml [hunk 10]
```diff
@@ -1,36 +1,27 @@
...
+#   - AWS_S3_TERRAFORM_STATE_BUCKET_NAME: S3 bucket for Terraform state
```
Comment (satsie) [inline .github/workflows/ci.yml]: Same here. This part of the documentation got removed:

> The S3 bucket name which is the output `terraform_state_bucket_name` from running the Terraform baseline infrastructure in `terraform/base`.

File: .github/workflows/ci.yml [hunk 11]
```diff
@@ -1,36 +1,27 @@
...
+#   - AWS_S3_TERRAFORM_STATE_OBJECT_KEY: State file key in the bucket
```
Comment (satsie) [inline .github/workflows/ci.yml]: More verbose but I think this description from the old documentation is good:

> The name of the *file* which will hold the Terraform state inside the above bucket.

File: .github/workflows/ci.yml [hunk 12]
```diff
@@ -1,36 +1,27 @@
...
+#   - ENVIRONMENT (optional): Environment name for namespacing
+#   - HOSTNAME: Application hostname (ex. api.savingsatoshi.com)
+#   - MAX_SCRIPT_EXECUTION_TIME (optional): REPL timeout in ms (default: 10000)
+#   - PORT (optional): Server port (default: 8000)
+#   - WHITELIST (optional): CORS allowed origins
```
Comment (satsie) [inline .github/workflows/ci.yml]: Suggestion: add back this line from the old documentation, "a comma-separated list"

File: src/routes/v1/status.ts [hunk 13]
```diff
@@ -1,11 +1,35 @@
...
-    status: 'OK',
-  })
+/**
+ * Health check endpoint for ALB target group health checks
+ * Verifies database connectivity to ensure the instance is ready to serve traffic
+ */
+router.get('/', async (req, res) => {
+  try {
+    // Verify database connectivity with a simple query
+    await prismaClient.$queryRaw`SELECT 1`
```
Comment (satsie) [inline src/routes/v1/status.ts:13]: This is a great idea!

File: terraform/scaling.tf [hunk 14]
```diff
@@ -0,0 +1,265 @@
...
+  step_adjustment {
+    scaling_adjustment          = 2
+    metric_interval_lower_bound = 0
+    metric_interval_upper_bound = 20
+  }
+
+  # 95%+ CPU: add 4 instances (20+ above threshold)
+  step_adjustment {
+    scaling_adjustment          = 4
+    metric_interval_lower_bound = 20
```
Comment (satsie) [inline terraform/scaling.tf:75]: How does this configuration work for 95%+ CPU? Why isn't `metric_interval_lower_bound` set to zero like in the `step_adjustment` above for 75-95%? I guess I'm just failing to see how setting something to 20 can correspond to 95% of something 😆

File: ansible/README.md [hunk 15]
```diff
@@ -1,38 +1,57 @@
...
   hosts:
-    # This IP should be the output `ip_address` from `terraform-apply`. It should be the EIP
-    # of your provisioned EC2 instance.
-    XXX.XXX.XXX.XX:
+    XXX.XXX.XXX.XX:           # Public IP of your VPS
       ansible_user: ubuntu
-      # This file path should be to the SSH key that has access to the instance.
-      ansible_ssh_private_key_file: /home/satoshi/saving-satoshi/saving-satoshi-test.pem
-      aws_region: us-west-2
-      # This `hostname` should be that which has a DNS A record pointed to the instance EIP.
```
Comment (satsie) [inline ansible/README.md:22]: Does this comment about `hostname` no longer apply?

**Review by satsie — APPROVED** [review]
Really good stuff! It's great to have a look at the code after seeing your very interesting and informative demo last week. Nearly all of my comments are just clarifying questions. A few things on some lines of documentation that were deleted. 

Agree with the decision to leave `ansible/playbook.yaml` in there. It might be more to maintain but I don't anticipate changes to the IaC that you have set up anytime soon. This is a really nice backup, especially for the reasons you describe in the documentation.

> Rename and expose the fallback_instance_types variable through GitHub Actions to allow specification of the ASG backup instance types.

yay! Thank you for this. So if we wanted to always enforce a c6a.xlarge instance type, we'd have that be the only one in the array?

> I also ran the tests on a c6a.xlarge instance today out of curiousity, and it performed very well. Not perfect, but very well.

That's great to hear. I looked up the pricing and t3.small is about $0.02/hour while c6a.xlarge is about $0.15. Is the recommendation to bump the instance up to c6a.xlarge before we know the system is going to be getting a lot of traffic (i.e. a workshop or something like the BOSS Challenge), and then bring it back down when it is over? If that is the case, maybe c6a.xlarge shouldn't be the default in `variables.pkrvars.hcl`? (unless I am misunderstanding how that variable works)

image

Spectacular work and documentation as usual. Thanks for making this idiot-proof! I'm happy to see this merged whenever you feel comfortable. I talked to Jamal earlier this week and he has been testing this PR on his personal AWS account. As of our conversation on Wednesday it was all working smoothly but let's wait for him to report back. Next week we can do some testing on staging. Exciting!

File: terraform/alb.tf [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+  }
+
+  # Sticky sessions for WebSocket connections
+  # WebSocket connections are inherently sticky once upgraded,
+  # but this helps with reconnection scenarios
+  stickiness {
+    type            = "lb_cookie"
+    cookie_duration = 86400 # 24 hours
+    enabled         = true
+  }
```
Comment (jrakibi) [inline terraform/alb.tf:143]: is that what assures us each student will always stay in the same EC2 instance? cause we probably gonna need students to stay on the same instance when using WebSocket

File: terraform/alb.tf [hunk 17]
```diff
@@ -0,0 +1,222 @@
...
+  validation_method = "DNS"
+
+  lifecycle {
+    create_before_destroy = true
+  }
+
+  tags = {
+    Name = "${local.namespace}-cert"
+  }
+}
```
Comment (jrakibi) [inline terraform/alb.tf:19]: are these ACM certificates free? we don't get charged for them, right? asking cause we'd issue a new certificate each time we push to develop, since the staging infra gets destroyed every night fromm the cron job.

File: terraform/asg.tf [hunk 18]
```diff
@@ -0,0 +1,317 @@
...
+}
+
+resource "aws_security_group_rule" "app_ssh" {
+  type              = "ingress"
+  from_port         = 22
+  to_port           = 22
+  protocol          = "tcp"
+  cidr_blocks       = ["0.0.0.0/0"]
+  security_group_id = aws_security_group.app.id
+  description       = "Allow SSH access (consider restricting in production)"
```
Comment (jrakibi) [inline terraform/asg.tf:45]: should we restrict this port? SSH is currently open to 0.0.0.0/0

Screenshot 2026-05-23 at 19 52 25

File: ansible/saving-satoshi-backend-alb.nginx.j2 [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+    listen [::]:80 default_server;
+    server_name _;
+
+    # Health check endpoint for ALB target group
+    # Returns 200 OK for ALB health checks
+    location /health {
+        access_log off;
+        return 200 "healthy\n";
+        add_header Content-Type text/plain;
+    }
```
Comment (jrakibi) [inline ansible/saving-satoshi-backend-alb.nginx.j2:16]: I think we can remove this block, the actual ALB health check is `/v1/status`, this just returns a hardcoded "healthy" string.

_logs from my console:_

```
saving-satoshi-backend-1 % curl -s https://api.ss.jrakibi.com/health        
healthy

saving-satoshi-backend-1 % curl -s https://api.ss.jrakibi.com/v1/status                       
{"status":"OK","timestamp":"2026-05-23T10:01:56.036Z","checks":{"database":"healthy"}}%       

```
Image

File: DEPLOYMENT.md [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+The backend runs on AWS. Pushing to `master` or `develop` automatically deploys the app — no manual steps needed after initial setup. A load balancer distributes incoming requests across one or more servers, and the server count scales up automatically under heavy traffic and back down when things are quiet.
+
+## Environments
+
+| Branch | Environment | Notes |
+|--------|-------------|-------|
+| `master` | Production | Persistent — never torn down |
+| `develop` | Staging | Infrastructure is destroyed nightly and rebuilt on the next push |
+
+Staging auto-destruction runs nightly at 4am UTC. This keeps costs near zero when staging is idle.
```
Comment (jrakibi) [inline DEPLOYMENT.md:12]: Not sure if there’s something wrong with my setup, but I ran the workflow on Wednesday: https://github.com/jrakibi/saving-satoshi-backend/actions/runs/26144416728. it’s Saturday now, and all the instances on my AWS account are still running. Are they supposed to be destroyed automatically?

File: .github/workflows/ci.yml [hunk 21]
```diff
@@ -146,25 +161,135 @@ jobs:
...
+        with:
+          filters: |
+            ami:
+              - 'packer/**'
+              - 'ansible/playbook-base.yaml'
+              - 'ansible/playbook-runtime.yaml'
+              - 'ansible/requirements.yaml'
+              - 'ansible/saving-satoshi-backend-alb.nginx.j2'
+              - 'ansible/saving-satoshi-backend.service.j2'
+              - 'src/base_images/**'
```
Comment (jrakibi) [inline .github/workflows/ci.yml:197]: FYI, I noticed a fresh deploy with an empty commit (git commit --allow-empty) triggers a full AMI rebuild even though no AMI affecting files changed. 

Here's the CI run where I noticed it: https://github.com/jrakibi/saving-satoshi-backend/actions/runs/26330696667. I was expecting it to skip the `Build AMI` job. (Prbly not worth fixing, just wanted to flag)

**Review by jrakibi — COMMENTED** [review]
Thanks for putting this together, @s373nZ. big PR, and I learned a lot reading through it.

I deployed it on my own AWS account from my fork, ran the full lifecycle (`deploy -> destroy -> redeploy`), and everything worked.

I also ran `make test-perf-load` against staging on a `t3.micro` to check the scale-up behavior. A few observations:

- Scale-up triggered ~5 min after sustained load began (1 → 2 instances)
- Each new instance took ~3 min from launch to serving traffic
- It got stuck at 2 instances. (I’ll probably run another test with `t3.small` to see if it matches your numbers)
- I ran `terraform destroy` manually to clean up the resources because it didn't run automatically from the 4am cron job as I was expecting (more on that in the review comments). The cleanup worked properly and took ~3 min.

Test log attached: [test-summary-t3.micro.log](https://github.com/user-attachments/files/28175413/test-summary-t3.micro.log)

CI run: https://github.com/jrakibi/saving-satoshi-backend/actions/runs/26330696667 

I left some comments below. Most are questions or small things. Nothing blocking the merge.

File: packer/variables.pkrvars.hcl (hunk 3 above)
Comment (jrakibi) [reply packer/variables.pkrvars.hcl:5]: I guess `c6a.xlarge` is just a temporary EC2 instance that Packer uses to create the AMI and install things on it.
As far as I understand, an AMI doesn't have an instance_type attached to it. same AMI can be used to launch any instance size.

Comment (s373nZ) [reply packer/variables.pkrvars.hcl:5]: Yes, @jrakibi is right. We are using the larger instance in this case temporarily, only to create the AMI. It should run for 5-10 minutes or so. The idea is to shave some time off of the AMI step in the pipeline.

File: packer/variables.pkrvars.hcl (hunk 4 above)
Comment (s373nZ) [reply packer/variables.pkrvars.hcl:8]: Yes, this is the place to update the base AMI used to build our custom AMI to Ubuntu `26.04` when you're ready. However, at that point, we'll want to make sure the Ansible steps are still compatible with the new version of Ubuntu -- sometimes new package versions can break things. Just saying there's a little more to thinking about when planning for the upgrade.

File: ansible/playbook-runtime.yaml (hunk 6 above)
Comment (s373nZ) [reply ansible/playbook-runtime.yaml:4]: These variables are defined between lines #43-#51 of `terraform/templates/user-data.sh`. This is the shell script the EC2 instance runs when it boots up, which in turn, runs Ansible. So the variables are setup there, and some are fed in through the Terraform.

You can see in that file that `APP_ENV_CONTENT` is being constructed by a combination of other variables, some of which come from GitHub Actions. Hope that helps -- let me know if you have more questions about this.

File: ansible/playbook-runtime.yaml (hunk 7 above)
Comment (s373nZ) [reply ansible/playbook-runtime.yaml:87]: Good question! It does. I was questioning this myself during the demo. But I think that we're doing it in both places just to be sure a strange bug doesn't make the REPLs unusable if they are not built into the AMI correctly.

They should be the same, and the run in this file should be cached by Docker... but if it's not, it means the new app server will just rebuild the REPL images without needing special intervention to fix the bug and go through the whole AMI process.

File: .env.example (hunk 8 above)
Comment (s373nZ) [reply .env.example:12]: Maybe just needs a modification to the `DATABASE_URL` setting in the GitHub Actions variable?

File: ansible/README.md (hunk 15 above)
Comment (s373nZ) [reply ansible/README.md:22]: Not really. As we're now using Route53 to manage DNS, a record is created automatically by the Terraform stack, mapping it to the ALB. We don't have an EIP anymore, just the ALB DNS name.

File: packer/ami.pkr.hcl (hunk 2 above)
Comment (s373nZ) [reply packer/ami.pkr.hcl:23]: Yes it is. If you want to move regions, you'll need to provision a new AMI the new region and use those AMIs everywhere. Migrating regions is inconvenient, but possible - you would need to update the hardcoded values and probably a lot of the GitHub Actions vars and code. Supporting multi-region is another level of complexity entirely.

File: terraform/alb.tf (hunk 16 above)
Comment (s373nZ) [reply terraform/alb.tf:143]: Yes sir! Sticky session configuration keep a user session tied to the same application instance.

File: terraform/alb.tf (hunk 17 above)
Comment (s373nZ) [reply terraform/alb.tf:19]: Yes, ACM certificates are free. They are provisioned with Terraform, so we re-use the same one across code pushes. It is tied to the ALB and the TLS is happening there. No new certs when code is pushed.

File: ansible/saving-satoshi-backend-alb.nginx.j2 (hunk 19 above)
Comment (s373nZ) [reply ansible/saving-satoshi-backend-alb.nginx.j2:16]: I'm 90% sure you're right, but I'm a little paranoid about merging this without going through another full test. Do you think it's ok to make a follow up issue for this?

File: DEPLOYMENT.md (hunk 20 above)
Comment (s373nZ) [reply DEPLOYMENT.md:12]: I took a quick look at your fork. I don't think your repo's "main" branch has the Staging Cleanup CI code in it? Maybe you either need to sync your fork with the latest, or set your repo's main branch to `develop` and make sure this PR code is also on your `develop`?

File: .github/workflows/ci.yml (hunk 21 above)
Comment (s373nZ) [reply .github/workflows/ci.yml:197]: Interesting... I haven't used the `--allow-empty` flag before, but it looks from the CI run that it's representing the full changeset between `master` and `develop`? Sound like that would mean it thinks the change set is *every* file in this PR?

```
Change detection refs/remotes/origin/master...refs/remotes/origin/develop
/usr/bin/git diff --no-renames --name-status -z refs/remotes/origin/master...refs/remotes/origin/develop
  M.env.exampleD.github/workflows/check-all-test.ymlA.github/workflows/ci.ymlD.github/workflows/deploy.ymlA.github/workflows/staging-destroy.yamlM.gitignoreADEPLOYMENT.mdMMakefileMREADME.mdMansible/README.mdAansible/playbook-base.yamlAansible/playbook-runtime.yamlMansible/requirements.yamlAansible/saving-satoshi-backend-alb.nginx.j2Mpackage.jsonApacker/.gitignoreApacker/ami.pkr.hclApacker/variables.pkrvars.hclMsrc/index.tsMsrc/routes/v1/status.tsAterraform/alb.tfAterraform/asg.tfAterraform/data.tfMterraform/ec2.tfMterraform/locals.tfAterraform/scaling.tfAterraform/ssm.tfAterraform/templates/user-data.shMterraform/variables.tfAtest/performance/README.mdAtest/performance/processors/functions.jsAtest/performance/profiles/http-tests.yamlAtest/performance/profiles/repl-capacity.yaml
```

I'm not sure, but feels like this is an edge case for the `paths-filter` CI step?

**Comment by s373nZ** [comment]
Added a commit to restore some of the documentation snippets per feedback.

File: terraform/scaling.tf (hunk 14 above)
Comment (s373nZ) [reply terraform/scaling.tf:75]: Yeah, this is tricky for me to understand too. Claude's explanation is better than mine:
> The bounds in step_adjustment are not absolute CPU percentages — they're offsets relative to the alarm threshold. The alarm threshold is 75 (line 87). AWS evaluates each step as: "how far is the current metric
  above the threshold?"
>
>  So:
>  - Step 1 — lower_bound = 0, upper_bound = 20: fires when CPU is 0–20 points above 75%, i.e. 75–95% → add 2 instances
>  - Step 2 — lower_bound = 20, no upper bound: fires when CPU is 20+ points above 75%, i.e. 95%+ → add 4 instances
>
>  The reason step 2's lower bound is 20 and not 0 is to avoid overlap with step 1. Each step must pick up exactly where the previous one left off. If both had lower_bound = 0, they'd be ambiguous and AWS would reject the policy at apply time.

File: ansible/saving-satoshi-backend-alb.nginx.j2 (hunk 5 above)
Comment (s373nZ) [reply ansible/saving-satoshi-backend-alb.nginx.j2:3]: This is a good question. It's port `80` because the ALB has the TLS certificate attached to it and is doing SSL termination, then forwarding traffic on to the application instance on port 80. Your application server is in a VPC, the default AWS VPC. However, it is in a public subnet, meaning it's accessible to the larger internet. As such, someone who found the IP address or DNS name of the instance could access the API on port `80`, which is not great.

The normal, secure pattern is to provision the EC2 instance into a *private* subnet, not accessible to the internet, and allow the ALB to bridge public traffic into it. This requires doing a more advanced VPC setup, and probably some more cost. Definitely something worth tracking in a follow up issue or discussing internally.

There are tradeoffs. If the app is in a private VPC, you can't SSH to it like was being done before without more setup, so I see why it was done like this initially. However, we have removed TLS, so access via port `80` is possible, just hard to find. If you wanted to try and add the LetsEncrypt cert back into Ansible, then we might be doing double-SSL on the ALB and the instance, or need reconfiguring of the ALB.

This is obviously a loaded topic. Let me know here or on Discord if you have more questions.

cc @jrakibi

File: terraform/asg.tf (hunk 18 above)
Comment (s373nZ) [reply terraform/asg.tf:45]: Restricting SSH is always a good idea -- but to what? Do we have a whitelist of IP addresses that should be able to SSH? Or are we adventurous enough to lock ourselves out?

If you want to restrict, the best-practice solution is to close port `22` and setup SSM Session Manager on the instance and use the AWS Console or API calls to start a shell session.

https://docs.aws.amazon.com/systems-manager/latest/userguide/session-manager-working-with.html

Probably worth looking into in a follow up?

**Comment by s373nZ** [comment]
@satsie Thanks for the review!

> So if we wanted to always enforce a c6a.xlarge instance type, we'd have that be the only one in the array?

Yes, indeed. I've tested this on my fork.

> Is the recommendation to bump the instance up to c6a.xlarge before we know the system is going to be getting a lot of traffic (i.e. a workshop or something like the BOSS Challenge), and then bring it back down when it is over?

I'd say that's good place to start -- depending on the activity, you might need to go higher for the BOSS challenge, but will need to experiment and observe to find out!

File: ansible/saving-satoshi-backend-alb.nginx.j2 (hunk 5 above)
Comment (s373nZ) [reply ansible/saving-satoshi-backend-alb.nginx.j2:3]: Thinking about this again this morning and double checked the code... I think we're actually good - port 80 is only accessible by the ALB because the security group only allows access to that port from the ALB security group...

So no, the port isn't protected by a private VPC subnet, but by the security group rule. Scratch my earlier concern. Sorry for the confusion.

File: terraform/alb.tf (hunk 17 above)
Comment (jrakibi) [reply terraform/alb.tf:19]: Oooh, though the nightly job destroys the staging certificate too, so we have to generate fresh ones after each new deploy

File: terraform/asg.tf (hunk 18 above)
Comment (jrakibi) [reply terraform/asg.tf:45]: yeah makes sense, we can leave it as is for now and revisit if needed.

File: ansible/saving-satoshi-backend-alb.nginx.j2 (hunk 19 above)
Comment (jrakibi) [reply ansible/saving-satoshi-backend-alb.nginx.j2:16]: I'll open a follow issue for that

File: .github/workflows/ci.yml (hunk 21 above)
Comment (jrakibi) [reply .github/workflows/ci.yml:197]: yeah just wanted to flag this. fine to leave as is

**Review by jrakibi — APPROVED** [review]
ACK fc85ff7dcefa91f84222ffdb8576226f3f144867 

I ran another test where I started the load test twice simultaneously, and the ASG scaled up to 5 instances.

Also, the 4 AM "Destroy Staging" cron job seems to be working fine now. I checked today and confirmed that the staging instances had been terminated/removed automatically.

Screenshot 2026-05-25 at 18 11 42

File: DEPLOYMENT.md (hunk 20 above)
Comment (jrakibi) [reply DEPLOYMENT.md:12]: Okay, I changed the develop branch to be the `default` branch for my fork, ran the `Destroy Staging` CI again, and now it seems to be skipping the Build AMI step as expected:
https://github.com/jrakibi/saving-satoshi-backend/actions/runs/26358433675

File: ansible/saving-satoshi-backend-alb.nginx.j2 (hunk 5 above)
Comment (satsie) [reply ansible/saving-satoshi-backend-alb.nginx.j2:3]: Oh very good! Thank you for looking into this. I didn't realize how complicated this could get!

File: ansible/playbook-runtime.yaml (hunk 6 above)
Comment (satsie) [reply ansible/playbook-runtime.yaml:4]: I totally missed that `APP_VERSION` was also being defined in `user-data.sh`. Makes sense now. Thank you for explaining!

File: terraform/scaling.tf (hunk 14 above)
Comment (satsie) [reply terraform/scaling.tf:75]: Wow, I guess that does make sense!

File: ansible/saving-satoshi-backend-alb.nginx.j2 (hunk 19 above)
Comment (satsie) [reply ansible/saving-satoshi-backend-alb.nginx.j2:16]: I opened this issue so it doesn't get lost: https://github.com/saving-satoshi/saving-satoshi-backend/issues/111
